### PR TITLE
Change SendFileAsync to use a fallback implementation instead of throwing

### PR DESCRIPTION
**Pull request by davidfowl**
- If the user wants to use the SendFile API directly then they can access the feature
explicitly.

#496

### CODE REVIEW

File: src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs [hunk 1]
```diff
@@ -74,10 +74,56 @@ public static Task SendFileAsync(this HttpResponse response, string fileName, lo
...
+        {
+            cancel.ThrowIfCancellationRequested();
+
+            if (string.IsNullOrWhiteSpace(fileName))
+            {
+                throw new ArgumentNullException(nameof(fileName));
+            }
+            if (!File.Exists(fileName))
+            {
+                throw new FileNotFoundException(string.Empty, fileName);
```
Comment (muratg) [inline src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs]: Why not just put `fileName` instead?

File: src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs [hunk 2]
```diff
@@ -33,7 +33,7 @@ public static bool SupportsSendFile(this HttpResponse response)
         /// Sends the given file using the SendFile extension.
         /// </summary>
         /// <param name="response"></param>
-        /// <param name="fileName"></param>
+        /// <param name="fileName">The full or relative path to the file.</param>
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs]: Callers should use the full path. Relative paths are unpredictable.

File: src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs [hunk 3]
```diff
@@ -74,10 +74,56 @@ public static Task SendFileAsync(this HttpResponse response, string fileName, lo
...
+                FileShare.ReadWrite,
+                bufferSize: 1024 * 64,
+                options: FileOptions.Asynchronous | FileOptions.SequentialScan);
+
+            try
+            {
+                fileStream.Seek(offset, SeekOrigin.Begin);
+
+                await StreamCopyOperation.CopyToAsync(fileStream, outputStream, length, cancel);
+            }
```
Comment (muratg) [inline src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs]: Should you catch instead and log?

Comment (davidfowl) [inline src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs]: Log with what?

File: src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs [hunk 4]
```diff
@@ -74,10 +74,56 @@ public static Task SendFileAsync(this HttpResponse response, string fileName, lo
...
         }
+
+        // Not safe for overlapped writes.
+        private static async Task SendFileAsync(Stream outputStream, string fileName, long offset, long? length, CancellationToken cancel)
+        {
+            cancel.ThrowIfCancellationRequested();
+
+            if (string.IsNullOrWhiteSpace(fileName))
+            {
+                throw new ArgumentNullException(nameof(fileName));
```
Comment (Eilon) [inline src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs]: Wrong exception type. Can this throw earlier, where the parameter is provided?

File: src/Microsoft.AspNet.Http.Extensions/StreamCopyOperation.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+{
+    // FYI: In most cases the source will be a FileStream and the destination will be to the network.
+    internal static class StreamCopyOperation
+    {
+        private const int DefaultBufferSize = 1024 * 16;
+
+        internal static async Task CopyToAsync(Stream source, Stream destination, long? length, CancellationToken cancel)
+        {
+            long? bytesRemaining = length;
+            byte[] buffer = new byte[DefaultBufferSize];
```
Comment (muratg) [inline src/Microsoft.AspNet.Http.Extensions/StreamCopyOperation.cs]: nit: var

File: src/Microsoft.AspNet.Http.Extensions/StreamCopyOperation.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+            {
+                // The natural end of the range.
+                if (bytesRemaining.HasValue && bytesRemaining.Value <= 0)
+                {
+                    return;
+                }
+
+                cancel.ThrowIfCancellationRequested();
+
+                int readLength = buffer.Length;
```
Comment (muratg) [inline src/Microsoft.AspNet.Http.Extensions/StreamCopyOperation.cs]: nit: var

File: src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs [hunk 7]
```diff
@@ -74,10 +74,56 @@ public static Task SendFileAsync(this HttpResponse response, string fileName, lo
             var sendFile = response.HttpContext.Features.Get<IHttpSendFileFeature>();
             if (sendFile == null)
             {
-                throw new NotSupportedException(Resources.Exception_SendFileNotSupported);
+                return SendFileAsync(response.Body, fileName, offset, count, cancellationToken);
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs]: This fallback makes SupportsSendFile unnecessary, remove it.

Comment (davidfowl) [inline src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs]: It's not unnecessary. It's just convenient. But sure.

File: src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs [hunk 8]
```diff
@@ -74,10 +74,56 @@ public static Task SendFileAsync(this HttpResponse response, string fileName, lo
...
+            {
+                throw new ArgumentOutOfRangeException(nameof(length), length, string.Empty);
+            }
+
+            var fileStream = new FileStream(
+                fileName,
+                FileMode.Open,
+                FileAccess.Read,
+                FileShare.ReadWrite,
+                bufferSize: 1024 * 64,
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs]: This is a 64kb buffer, but StreamCopyOperation uses a 16kb buffer. Would it help if they were aligned? Do we know if this is a managed or native buffer?

Comment (davidfowl) [inline src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs]: Copied your codez bro :smile: . This should totally use the memory pool.

File: src/Microsoft.AspNet.Http.Extensions/StreamCopyOperation.cs (hunk 5 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.Http.Extensions/StreamCopyOperation.cs]: https://github.com/aspnet/StaticFiles/issues/91 needs to move to this repo as well.

File: src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs (hunk 3 above)
Comment (muratg) [inline src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs]: Ugh just thought there would be a logger here. Apparently not.

File: src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs (hunk 4 above)
Comment (davidfowl) [inline src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs]: Copy pasta

File: src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs (hunk 8 above)
Comment (davidfowl) [inline src/Microsoft.AspNet.Http.Extensions/SendFileResponseExtensions.cs]: @Tratcher Maybe they should be the same.

File: src/Microsoft.AspNet.Http.Extensions/StreamCopyOperation.cs (hunk 5 above)
Comment (davidfowl) [inline src/Microsoft.AspNet.Http.Extensions/StreamCopyOperation.cs]: Use the fancy issue mover

**Pull request closed without merging**